pep-0682.rst

```diff
@@ -0,0 +1,217 @@
+PEP: 682
+Title: Format Specifier for Signed Zero
+Author: John Belmonte <john@neggie.net>
+Sponsor: Mark Dickinson <dickinsm@gmail.com>
+Discussions-To:
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 29-Jan-2022
+Python-Version: 3.11
+Post-History:
+
+
+Abstract
+========
+
+Though ``float`` and ``Decimal`` types can represent `signed zero`_, in many 
+fields of mathematics negative zero is surprising or unwanted -- especially
+in the context of displaying an (often rounded) numerical result.  This PEP
+proposes an extension to the `string format specification`_ allowing negative
+zero to be normalized to positive zero.
+
+.. _`signed zero`: https://en.wikipedia.org/wiki/Signed_zero
+.. _`string format specification`: https://docs.python.org/3/library/string.html#formatstrings
+
+
+Motivation
+==========
+
+Here is negative zero:
+
+.. code-block:: pycon
+
+    >>> x = -0.
+    >>> x
+    -0.0
+
+When formatting a number, negative zero can result from rounding.  Assuming
+the user's intention is truly to discard precision, the distinction between
+negative and positive zero of the rounded result might be considered an
+unwanted artifact:
+
+.. code-block:: pycon
+
+    >>> for x in (.002, -.001, .060):
+    ...     print(f'{x: .1f}')
+     0.0
+    -0.0
+     0.1
+
+There are various approaches to clearing the sign of a negative zero.  It
+can be achieved without a conditional by adding positive zero:
+
+.. code-block:: pycon
+
+    >>> x = -0.
+    >>> x + 0.
+    0.0
+
+To normalize negative zero when formatting, it is necessary to perform
+a redundant (and error-prone) pre-rounding of the input:
+
+.. code-block:: pycon
+
+    >>> for x in (.002, -.001, .060):
+    ...     print(f'{round(x, 1) + 0.: .1f}')
+     0.0
+     0.0
+     0.1
```

Comment:
By the way, when `code-block:: pycon` is used, I noticed that the official rendering applies a proportional font to `>>>` and `...` specifically, which looks rather odd and affects indent alignment.

![Screen Shot 2022-02-09 at 8 35 43 AM](https://user-images.githubusercontent.com/1708631/153093888-ab42c4e5-0c74-4ed3-99a8-9f8a0fb4abb2.png)

Reply:
Hmm, this isn't ideal. It's a pythondotorg styling problem, I'd imagine - and hard to fix currently.

A

Reply:
Grep shows only one other PEP using `code-block:: pycon`, so there's no safety in numbers.  I'll remove the usage for now, as part of #2317.

Reply:
Why would it be "pycon", and not "python"? It seems like all of the other PEPs don't specify a language, so I think you're right in removing it.

Reply:
> Why would it be "pycon", and not "python"

`pycon` is the Pygments lexer for python console text, versus `python` which is for scripts. Syntactically `pycon` is correct here, although I don't know the exact difference between the two.

> It seems like all of the other PEPs don't specify a language

Support for specifying languages was only added very recently (by adding Pygments to the requirements file) -- as a PEP editor I would strongly encourage explicitness in language as it makes reading PEPs easier with correctly highlighted code.

A

Reply:
(`pycon` main differences seem to be that it doesn't highlight console output as Python code, and also highlights tracebacks)

Reply:
A bit late now, sorry (I was invited to visit the SpaceX Starship production and launch site in Texas and thus was unavailable for the past week) but I can fully confirm everything @AA-Turner said, which was the basis of my initial implemented suggestion. In addition to the reasons he mentioned, being explicit about the language in the code block is also ensures the block will be highlighted correct regardless of what we decide to do in terms of the default syntax highlighter (none, auto, Python, etc).

Also, I cannot reproduce the reported issue on Firefox/Windows the [deployed PEP](https://www.python.org/dev/peps/pep-0682/) on Python.org rendered via the legacy docutils build system. If you're seeing it there, this is likely due to platform/OS/version-specific font/rendering differences given the current "system font stack" approach to monospace fonts, which is why I advocate using a webfont instead to guarantee consistent, correct rendering on all platforms, OS versions and user configurations to avoid a plethora of possible edge-case display issues like this with the high number of possible combinations of monospace, sans and serif fonts that could be selected depending on the OS, version, browser and user config. NB @AA-Turner @hugovk

Reply:
Thank you-- while we did revert the pycon block annotations for now, I agree that correct block annotation is desirable and we'd like all the PEP documents to be doing that.

Summary from the time of the decision (https://github.com/python/peps/pull/2317#discussion_r804252471):
to summarize:
  * the console-specific formatting looks great with the new rendering system for peps.python.org
  * however, rendering on the standard pep site is botched.  @AA-Turner notes it may be hard to fix.  (I'd be happy to help if it's feasible.)
  * there is only one other PEP so far using `pycon`
  * it seems we'll need a script to convert console blocks in existing PEP's anyway, so perhaps no need to put this PEP on the bleeding edge

Reply:
Yes, but as I noted there, all this is specific to the `pycon` highlighter; simply using the `python` highlighter instead should fix the issue (unless you've confirmed it doesn't, but I didn't find any mention of that here nor there). And I am not aware of plans for batch-converting all old PEPs to use the new explicit language markup, except in cases on active/process PEPs where the highlighting is widely off (wrong language, etc) but there's no reason not to be explicit and correct in new ones as @AA-Turner says, unless it is not compatible with both renderers.

Reply:
> batch-converting all old PEPs to use the new explicit language markup

I have avoided this topic as I don't want to spark a discussion. My current opinion is we should leave implicit literal block syntax as it is in most places, but update all cases where the language is not Python. This opinion is subject to change several times per arbitrary time period.

This probably means we wouldn't update to `pycon` here on a retroactive sweep. 

> typefaces

another debate for another time ;)

A

Reply:
> I have avoided this topic as I don't want to spark a discussion. My current opinion is we should leave implicit literal block syntax as it is in most places, but update all cases where the language is not Python. This opinion is subject to change several times per arbitrary time period.

I share this opinion as well, and am willing to help implement if and when the time comes for this, though I think it could be limited to not include withdrawn, superceded, and otherwise inactive PEPs. For new PEPs I would favor the `.. code-block:: python` syntax for all code blocks, which I believe is your viewpoint as well.